### PR TITLE
Implement non degradation comparisons

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Issue #6 

This PR adds the first half of the comparisons listed in the issue.

# Changes/Additions

1. Add the `comparison_generator` function to generate random comparisons without cluttering the `random_plot_generator`.

### CODE REVIEW

File: bot/plotting/comparison_generator.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    comparison_dict = {}
+
+    if number_of_comp == 1:
+        param_list = []
+        diff_params = random.randint(2, 3)
+        for i in range(0, diff_params):
+            param_list.append(params.copy())
+            param_list[i][
+                "Current function [A]"
+            ] = chemistry_generator(
```
Comment (rtimms) [inline bot/plotting/comparison_generator.py:25]: Not for this issue, but in `chemistry_generator` can we just take the default value from the `chemistry` and then pick a number in some range based on that rather than having to specify an interval to pick from for each parameter set? I guess the range might be different for different parameters?

Comment (Saransh-cpp) [reply bot/plotting/comparison_generator.py:25]: Yes, I was about to update and rename it to generate values in the +- 10% (except for `"Current function [A]"`) range - 
```py
        params = pybamm.ParameterValues(chemistry=chemistry)
        param_value = desired_decimal_point_generator(
            params[parameter]*0.9, params[parameter]*1.1, 2
        )
```
I will push the code

File: bot/plotting/comparison_generator.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        choice = provided_choice
+
+    if choice == 0:
+        s = pybamm.BatchStudy(
+            models=models_for_comp,
+            parameter_values=parameter_values_for_comp,
+            permutations=True,
+        )
+
+        s.solve([0, 3700])
```
Comment (rtimms) [inline bot/plotting/comparison_generator.py:43]: If you change the "Current function [A]" you might also want to change the amount of time you are solving for. The default value discharge in around 1 hour, so if e.g. you double the current you might only want to solve for around half an hour. An alternative would be to set it up as an experiment: 

```python3
model = pybamm.lithium_ion.SPM()

parameter_values = pybamm.ParameterValues(chemistry=pybamm.parameter_sets.Marquis2019)
V_low = parameter_values["Lower voltage cut-off [V]"]
I_1C = parameter_values["Current function [A]"]

factor = 2 # randomize this 
I = factor * I
t_end = 1 / factor * 1.1 

experiment = pybamm.Experiment([f"Discharge at {I} A for {t_end} hours or until {V_low}V"])
sim = pybamm.Simulation(model, experiment=experiment, parameter_values=parameter_values)
sim.solve()
```

This is a little slower than just doing a simulation without the experiment, so you could just alter the solve time in what you are already doing.

File: bot/plotting/comparison_generator.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                cycle = experiment_generator()
+                number = random.randint(1, 3)
+
+                if provided_choice is not None:
+                    experiment = [
+                        (
+                            "Discharge at C/10 for 10 hours "
+                            + "or until 3.3 V",
+                            "Rest for 1 hour",
+                            "Charge at 1 A until 4.1 V",
```
Comment (rtimms) [inline bot/plotting/comparison_generator.py:67]: Similarly here you might want to get the voltage limits from the parameter set. You might find it crashes if you try to go to too high/low voltage

**Review by rtimms — CHANGES_REQUESTED** [review]
Thanks, @Saransh-cpp looking good! Could you add some comments in the code as I'm finding it a little hard to follow? E.g. it would be good to say what `provided_choice` means.

**Comment by Saransh-cpp** [comment]
> Thanks, @Saransh-cpp looking good! Could you add some comments in the code as I'm finding it a little hard to follow? E.g. it would be good to say what `provided_choice` means.

Yes, the code lacks comments, I will add them.

File: bot/plotting/comparison_generator.py (hunk 2 above)
Comment (Saransh-cpp) [reply bot/plotting/comparison_generator.py:43]:
```py
factor = 2 # randomize this 
I = factor * I
t_end = 1 / factor * 1.1 
```
In this script, is `I` the default `"Current function [A]"` value? I think adding this script should make it work

Comment (rtimms) [reply bot/plotting/comparison_generator.py:43]: Yes, sorry that should be `I = factor * I_1C`

Comment (rtimms) [reply bot/plotting/comparison_generator.py:43]: where `I_1C` is the default value

Comment (Saransh-cpp) [reply bot/plotting/comparison_generator.py:43]: How should I go about providing the same t_eval for 2-3 different values of `"Current function [A]"` (which is happening here), should I take the max value as a reference or the average values as the reference?

Comment (rtimms) [reply bot/plotting/comparison_generator.py:43]: So here are you comparing the same model/parameters and just changing the "Current function [A]"? The smallest current will give you the longest time, so use that the calculate the simulation times. The other simulations with a higher current will just stop sooner because they will hit an event (e.g. the low voltage cut-off).

Comment (Saransh-cpp) [reply bot/plotting/comparison_generator.py:43]: Yes, that is what I am doing here. Thanks, I will add this!

File: bot/plotting/random_plot_generator.py [hunk 4]
```diff
@@ -95,37 +96,55 @@ def random_plot_generator(
...
-                pybamm.lithium_ion.SPMe(
-                    options=options
-                ),
+            solvers = [
+                pybamm.CasadiSolver(mode="safe"),
+                pybamm.CasadiSolver(mode="fast with events")
             ]
 
-            model = random.choice(models)
+            solver = random.choice(solvers)
```
Comment (brosaplanella) [inline bot/plotting/random_plot_generator.py:104]: Why do you choose the solver at random?

File: bot/plotting/random_plot_generator.py [hunk 5]
```diff
@@ -95,37 +96,55 @@ def random_plot_generator(
...
+
+            choice = random.randint(0, 3)
```
Comment (brosaplanella) [inline bot/plotting/random_plot_generator.py:106]: Using `choice` as an integer to distinguish between cases is not very informative. Are we set on 2 cases or we might get an arbitrary number in the future?

File: bot/plotting/random_plot_generator.py [hunk 6]
```diff
@@ -95,37 +96,55 @@ def random_plot_generator(
...
-            choice = random.randint(0, 2)
             if testing is True and provided_choice is not None:
                 choice = provided_choice
 
-            solvers = [
-                pybamm.CasadiSolver(mode="safe"),
-                pybamm.CasadiSolver(mode="fast"),
-                pybamm.CasadiSolver(mode="fast with events")
-            ]
+            if choice == 1:
```
Comment (brosaplanella) [inline bot/plotting/random_plot_generator.py:111]: For example, if the `choice` is to choose between passing options or default, I think it should use if you always pass options and the choice is between passing an option dictionary or `None` (which should fall back to the default).

File: test/test_comparsion_generator.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
+import unittest
```
Comment (brosaplanella) [inline test/test_comparsion_generator.py:1]: There is a typo in the file name (it says comparsion)

**Review by brosaplanella — CHANGES_REQUESTED** [review]
Looks good, just suggested some minor changes/comments.

File: bot/plotting/random_plot_generator.py (hunk 4 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:104]: No particular reason, just to experiment with the solvers. Should I just use the "safe" mode?

Comment (brosaplanella) [reply bot/plotting/random_plot_generator.py:104]: Technically they are both the same solver, with the difference that one can deal with events (safe) and the other one can't (fast). In general, you want to you fast when you don't have events and safe otherwise so typically it is not a choice. You could try other DAE solvers though if you want to compare them (e.g. Scipy DAEs I think?)

File: bot/plotting/random_plot_generator.py (hunk 5 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:128]: I think we might expand it in the future, I have added comments (pushing them) for what will happen if a possible random integer is selected, should I make it into a list with all the choices as elements and choose one of them randomly?

File: test/test_comparsion_generator.py (hunk 7 above)
Comment (Saransh-cpp) [reply test/test_comparsion_generator.py]: Yes, thanks!

File: bot/plotting/random_plot_generator.py (hunk 6 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py]: Oh yes, I've made the changes. Thanks!

File: bot/plotting/random_plot_generator.py (hunk 4 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:122]: That sounds interesting, I'll try setting it up on my windows machine. For now, should I just use the "safe" mode?

File: bot/plotting/random_plot_generator.py [hunk 8]
```diff
@@ -3,40 +3,50 @@
...
-    plot_summary_variables=True,
+    provided_chemistry=None,
     provided_degradation=True,
 ):
     """
     Generates a random plot.
     Parameters:
         return_dict: dict
+            A shared dictionary in which all the return values are stored.
         testing: bool
```
Comment (rtimms) [inline bot/plotting/random_plot_generator.py:25]: Just a suggestion, it might be cleaner to pass an `options` dict which is by defualt `{"choice": None, "number of comparisons": None, "chemistry": None}` then you can just check e.g. 
```python3
if options[chemistry] is None:
    options[chemistry] = random.choice(chemistries)
````

And you can do the same in the other scripts.

Comment (rtimms) [reply bot/plotting/random_plot_generator.py:25]: It probably shouldn't be called options to avoid clashing with the model options, or you could called the options for the model `model_options`

File: bot/plotting/random_plot_generator.py [hunk 9]
```diff
@@ -95,6 +113,28 @@ def random_plot_generator(
...
+            # 1: experiment with summary variable
+            # 2: experiment without summary variables
+            # 3: comparison plots
+            choice = random.randint(0, 3)
+
+            # if testing, don't randomly choose stuff
+            if testing is True and provided_choice is not None:
+                choice = provided_choice
+
+            # Add degradation only if we are plotting summary variables
```
Comment (rtimms) [inline bot/plotting/random_plot_generator.py:134]: you could move setting degradation options to here so that you don't necessarily set the options and then overwrite them

File: bot/plotting/random_plot_generator.py [hunk 10]
```diff
@@ -151,6 +184,7 @@ def random_plot_generator(
                     lower_voltage=lower_voltage,
                 )
 
+                # creating the GIF
                 time_array = plot_graph(solution, sim)
 
                 return_dict["model"] = model
```
Comment (rtimms) [inline bot/plotting/random_plot_generator.py:190]: might be cleaner to use `update({})`

File: bot/plotting/random_plot_generator.py [hunk 11]
```diff
@@ -166,45 +200,58 @@ def random_plot_generator(
                 return
 
             elif choice == 1:
+
+                # generating a random experiment
+                cycle_received = experiment_generator()
+                number = random.randint(4, 100)
```
Comment (rtimms) [inline bot/plotting/random_plot_generator.py:206]: very minor, but why not just let `experiment_generator` choose the number of repeats? Ah, I see that you want to pick a big/small number for long/short experiments.

File: bot/plotting/random_plot_generator.py [hunk 12]
```diff
@@ -95,6 +113,28 @@ def random_plot_generator(
...
+
+            # if testing, don't randomly choose stuff
+            if testing is True and provided_choice is not None:
+                choice = provided_choice
+
+            # Add degradation only if we are plotting summary variables
+            if choice != 1:
+                options = None
+
             models = [
```
Comment (rtimms) [inline bot/plotting/random_plot_generator.py:138]: Very minor point, but you could define 
```python3
            MODELS = [
                pybamm.lithium_ion.DFN(
                    options=options
                ),
                pybamm.lithium_ion.SPM(
                    options=options
                ),
                pybamm.lithium_ion.SPMe(
                    options=options
                ),
            ]
```
etc. at the top of the script. Then if we want to add more models, chemistries etc. you don't need to search the code for the list. Also, you can just see at the top of the script all the different combinations that might be compared. I'm not sure what is recommended style-wise?

**Review by rtimms — APPROVED** [review]
Thanks, @Saransh-cpp looks good! I've just made some suggestions

File: bot/plotting/random_plot_generator.py (hunk 11 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:206]: Yes, it was doing that before but to clean out the code I decided to shift the number out.

File: bot/plotting/random_plot_generator.py (hunk 8 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:25]: Yes that will make it more readable, I'll make the changes

File: bot/plotting/random_plot_generator.py (hunk 4 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:122]: Should I keep the "safe" mode for now? The "fast with events" mode sometimes results in the complete freezing of a simulation.

Comment (rtimms) [reply bot/plotting/random_plot_generator.py:122]: sounds good

File: bot/plotting/random_plot_generator.py (hunk 12 above)
Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:138]: For this, I have moved every such list outside the function but the `models` list depends upon the plot that we are creating if it a summary variable plot then the models get degradation otherwise not. Should I still move the `models` list out? and then redefine the list when we need to include degradation?

Comment (rtimms) [reply bot/plotting/random_plot_generator.py:138]: Could you do e.g. `MODELS = [pybamm.lithium_ion.SPM, pybamm.lithium_ion.DFN]` then create instances as necessary with options as  `a = models[i]()` and `b = models[i](options={"SEI":"ec reaction limited"})`?

Maybe it's better to leave this as it was since you end up setting different options depending on the plot and the parameter set anyway. Not sure of the best approach. Both are probably ok, so I'll leave it to you :)

Comment (Saransh-cpp) [reply bot/plotting/random_plot_generator.py:147]: I tried doing this but it didn't work out, I'll keep the present one for now. Thanks!

**Review by rtimms — APPROVED** [review]

**Comment by Saransh-cpp** [comment]
The last 3 commits were about adding some more tests and then realizing that the added tests were not passing due to Ai2020 parameter set and then fixing the same. The tests should pass once [#1521](https://github.com/pybamm-team/PyBaMM/pull/1521) is merged into PyBaMM develop.
